pkg/agent/multicluster/pod_route_controller.go

```diff
@@ -0,0 +1,394 @@
+// Copyright 2023 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package multicluster
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/util/wait"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/kubernetes"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	mcv1alpha1 "antrea.io/antrea/multicluster/apis/multicluster/v1alpha1"
+	mcclientset "antrea.io/antrea/multicluster/pkg/client/clientset/versioned"
+	"antrea.io/antrea/multicluster/pkg/client/informers/externalversions/multicluster/v1alpha1"
+	mclisters "antrea.io/antrea/multicluster/pkg/client/listers/multicluster/v1alpha1"
+	"antrea.io/antrea/pkg/agent/config"
+	"antrea.io/antrea/pkg/agent/interfacestore"
+	"antrea.io/antrea/pkg/agent/openflow"
+	"antrea.io/antrea/pkg/ovs/ovsconfig"
+)
+
+const (
+	// The number of workers processing a Pod change
+	podWorkerNum = 5
+
+	dummyKey               = "key"
+	podIndexKey            = "podIP"
+	podRouteControllerName = "MCPodRouteController"
+)
+
+// MCPodRouteController generates L3 forwarding flows to forward cross-cluster
+// traffic from MC Gateway to Pods on other Nodes inside a member cluster. It is
+// required when networkPolicyOnly, noEncap or hybrid mode are configured, to forward
+// the traffic through tunnels between Gateway and other Nodes, as otherwise the
+// traffic will not go through tunnels in those modes.
+type MCPodRouteController struct {
+	k8sClient       kubernetes.Interface
+	mcClient        mcclientset.Interface
+	ovsBridgeClient ovsconfig.OVSBridgeClient
+	ofClient        openflow.Client
+	interfaceStore  interfacestore.InterfaceStore
+	nodeConfig      *config.NodeConfig
+	podQueue        workqueue.RateLimitingInterface
+	gwQueue         workqueue.RateLimitingInterface
+	podInformer     cache.SharedIndexInformer
+	podLister       corelisters.PodLister
+	gwInformer      cache.SharedIndexInformer
+	gwLister        mclisters.GatewayLister
+	// podWorkersStarted is a boolean which tracks if the Pod flow controller has been started.
+	podWorkersStarted      bool
+	podWorkersStartedMutex sync.RWMutex
+	podWorkerStopCh        chan struct{}
```

Comment:
There are a few attributes never used, please clean them up and remove from the constructor.

Reply:
Removed, thanks.